TombEngine/Objects/TR3/Entity/tr3_civvy.cpp

```diff
@@ -79,341 +119,484 @@ namespace TEN::Entities::Creatures::TR3
+				case CIVVY_STATE_WALK:
 
-				else if (creature->Mood == MoodType::Escape)
-				{
-					if (Lara.TargetEntity != item && AI.ahead)
-						item->Animation.TargetState = CIVVY_STATE_IDLE;
-					else
-						item->Animation.TargetState = CIVVY_STATE_RUN_FORWARD;
-				}
-				else if (creature->Mood == MoodType::Bored ||
-					(item->AIBits & FOLLOW && (creature->ReachedGoal || laraAI.distance > SQUARE(SECTOR(2)))))
-				{
-					if (item->Animation.RequiredState != NO_STATE)
-						item->Animation.TargetState = item->Animation.RequiredState;
-					else if (AI.ahead)
-						item->Animation.TargetState = CIVVY_STATE_IDLE;
-					else
-						item->Animation.TargetState = CIVVY_STATE_RUN_FORWARD;
-				}
-				else if (AI.bite && AI.distance < CIVVY_ATTACK0_RANGE)
-					item->Animation.TargetState = CIVVY_AIM0;
-				else if (AI.bite && AI.distance < CIVVY_ATTACK1_RANGE)
-					item->Animation.TargetState = CIVVY_AIM1;
-				else if (AI.bite && AI.distance < CIVVY_WALK_RANGE)
-					item->Animation.TargetState = CIVVY_STATE_WALK_FORWARD;
-				else
-					item->Animation.TargetState = CIVVY_STATE_RUN_FORWARD;
-
-				break;
+					creature.MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
+					jointHeadRot.y = targetAngle;
 
-			case CIVVY_STATE_WALK_FORWARD:
-				creature->MaxTurn = CIVVY_WALK_TURN_RATE_MAX;
-				extraHeadRot.y = laraAI.angle;
-
-				if (item->AIBits & PATROL1)
-				{
-					extraHeadRot.y = 0;
-					item->Animation.TargetState = CIVVY_STATE_WALK_FORWARD;
-				}
-				else if (creature->Mood == MoodType::Escape)
-					item->Animation.TargetState = CIVVY_STATE_RUN_FORWARD;
-				else if (creature->Mood == MoodType::Bored)
-				{
-					if (Random::TestProbability(CIVVY_WAIT_CHANCE))
+					if (item.AIBits & PATROL1)
 					{
-						item->Animation.TargetState = CIVVY_STATE_IDLE;
-						item->Animation.RequiredState = CIVVY_WAIT;
+						jointHeadRot.y = 0;
+						item.Animation.TargetState = CIVVY_STATE_WALK;
 					}
-				}
-				else if (AI.bite && AI.distance < CIVVY_ATTACK0_RANGE)
-					item->Animation.TargetState = CIVVY_STATE_IDLE;
-				else if (AI.bite && AI.distance < CIVVY_ATTACK2_RANGE)
-					item->Animation.TargetState = CIVVY_AIM2;
-				else
-					item->Animation.TargetState = CIVVY_STATE_RUN_FORWARD;
+					else if (item.AIBits & FOLLOW)
+					{
+						if (creature.ReachedGoal)
+							item.Animation.TargetState = CIVVY_STATE_IDLE;
+						else if (targetDistance > CIVVY_WALK_RANGE)
+							item.Animation.TargetState = CIVVY_STATE_RUN;
+						else
+							item.Animation.TargetState = CIVVY_STATE_WALK;
+					}
+					else if (creature.Mood == MoodType::Escape)
+						item.Animation.TargetState = CIVVY_STATE_RUN;
+					else if (creature.Mood == MoodType::Bored)
+					{
+						if (Random::TestProbability(CIVVY_WAIT_CHANCE))
+							item.Animation.TargetState = CIVVY_STATE_IDLE;
+					}
+					else if ((creature.Enemy->HitPoints > 0 && creature.Enemy->Animation.Velocity.z < CIVVY_TARGET_ALERT_VELOCITY) &&
```

Comment:
Added the `creature.Enemy != nullptr` check.

The velocity.z is checking if the target (Lara or another enemy) is moving. If it isn't, then the civvy can stop (IDLE) and do the close punch. But if the target is moving, this rutine of stop and punch, is very likely to end hitting the air. If there were not other choice, this would be.

But, because this entity has another attack he can do on movement, I put that velocity.z check to make him decide, if the target is moving, don't stop and punch, instead do the walking punch (next else if).

Reply:
Okay approved :)